src/vm/costs/mod.rs

```diff
@@ -640,10 +641,14 @@ impl LimitedCostTracker {
             memory: 0,
             memory_limit: CLARITY_MEMORY_LIMIT,
             non_free: None,
-            mainnet: false,
+            mainnet: use_mainnet,
         }
     }
 
+    pub fn new_free() -> LimitedCostTracker {
+        LimitedCostTracker::new_free_on_network(false)
```

Comment:
I don't think this is right.  We use `LimitedCostTracker::new_free()` in mainnet-reachable code, such as in `src/net/rpc.rs`, in `src/vm/analysis/mod.rs` (in `type_check()`), and in `src/clarity_vm/clarity.rs` (i.e. in `begin_genesis_block()`).

Reply:
The free cost tracker never uses the mainnet flag — prior to this PR, I think the mainnet flag was always set to false. The cleanest way to have implemented the free tracker would probably to make the LimitedCostTracker an enum type, but that doesn’t end up having much to do with this PR.

Reply:
This feels unsafe and brittle.  Why should a cost tracker's `mainnet` flag be set to `false`, when the node is actually running on mainnet?

Currently, `LimitedCostTracker'`s mainnet flag is only used in calls to `boot_code_id()` and `load_cost_functions()` in `LimitedCostTracker::load_costs()`.  While it is the case that `LimitedCostTracker::load_costs()` is only called by `LimitedCostTracker::new()` and `LimitedCostTracker::new_mid_block()`, it seems that the only reason this doesn't cause immediate problems is because all of the `CostTracker` trait methods in the `LimitedCostTracker` implementation short-circuit if the `self.free` flag is set.  This short-circuiting prevents these trait method implementations from ever trying to access the loaded cost functions.  However, the act of trying to load a cost function in a `LimitedCostTracker` created by `new_free()` doesn't seem to panic; the trait method implementations would instead return `CostErrors::CostComputationFailed("Computation not defined: {}")`.  This, to me, looks like an accident waiting to happen -- for example, suppose we accidentally try to rely on `LimitedCostTracker`'s mainnet flag indicating the mainnet status of the system, like we did with `Burnchain`'s mainnet flag.

Is there a way we can de-risk this here without doing a lot of refactoring?  For example, could we make `load_costs()` panic if `self.free` is set, and make the `CostTracker` trait method implementations panic if `mainnet` and `free` are both true?

Reply:
The refactor is pretty straight-forward: https://github.com/blockstack/stacks-blockchain/pull/2945

Reply:
As I said on #2945 , this PR is meant for test coverage, but it introduces new changes to production code, which really should induce a new round of testing.

One option is to just *look* at the test coverage, but not actually merge this into the release, so that we don't introduce new changes to prod over the weekend.

Reply:
> This feels unsafe and brittle. Why should a cost tracker's mainnet flag be set to false, when the node is actually running on mainnet?

This new code is just making explicit the previous behavior. It's not changing the behavior, as kantai said. In other words, it was effectively using the value `mainnet: false` before.

Reply:
> One option is to just look at the test coverage, but not actually merge this into the release, so that we don't introduce new changes to prod over the weekend.

True, though I do believe we should merge the tests and refactor — this is useful to integrate into CI. Certainly doesn’t need to be merged on the weekend, though (and merging to next-costs by itself doesn’t mean production yet either).

Reply:
@kantai @jcnelson  If you guys want to merge it this weekend, go ahead.

Otherwise can defer.